cumulusci/utils/yaml/model_parser.py

```diff
@@ -0,0 +1,145 @@
+from typing import Union, IO
+from pathlib import Path, Sequence
+from typing_extensions import Literal
+
+from yaml import safe_load
+
+from pydantic import BaseModel, ValidationError, Field
+from pydantic.error_wrappers import ErrorWrapper
+
+from cumulusci.utils.fileutils import load_from_source
+
+_error_handling = "warn", "raise"
+ErrorHandling = Literal[_error_handling]
+
+
+class CCIModel(BaseModel):
+    """Base class with convenience features"""
+
+    _magic_fields = ["fields"]
+
+    @classmethod
+    def parse_from_yaml(cls, source: Union[str, Path, IO]):
+        path, data = load_from_source(source, safe_load)
+        return cls.parse_obj(data, path).__root__
+
+    @classmethod
+    def parse_obj(cls, data: [dict, list], path: str = None):
+        try:
+            return super().parse_obj(data)
+        except ValidationError as e:
+            _add_filenames(e, path)
+            raise e
+
+    @classmethod
+    def validate_data(
+        cls,
+        data: Union[dict, list],
+        context: str = None,
+        on_error: ErrorHandling = "raise",
```

Comment:
I also use typing_extensions and Literals in CumulusCI.yml for Source Format. There may be other places it would be appropriate to use in that schema as well.

I actually prefer strings to enums and now that Literals can be statically checked by MyPy, enums have very few advantages to balance out the extra typing verbosity (and the fact that they can't be represented in YAML or JSON).

Reply:
I still see some advantages to enums:
- it'll error even if you haven't configured your system to run the static type checker
- no need to add special libraries like typing_extensions that are only intended for particular versions of Python

And this particular case is not being loaded from YAML or JSON.

I think the case for strings instead of enums will be clearer once we drop support for Python < 3.7.

It also feels odd to use a Literal-typed string here but an enum in other locations:
- https://github.com/SFDO-Tooling/CumulusCI/blob/master/cumulusci/tasks/metadata_etl/base.py#L17
- https://github.com/SFDO-Tooling/CumulusCI/pull/1489/files#diff-b53c9e73bd178b56da320173d3534896R20

What are your thoughts/preferences on this, @davidmreed?

Reply:
We should discuss this further I agree.

But in this particular case, maybe this flag is vestigial.

There is no good reason for validate to throw an exception.

Reply:
https://github.com/SFDO-Tooling/CumulusCI/blob/59aac2203206dd75403e0eb629039886ff669bd1/cumulusci/utils/yaml/model_parser.py#L35